### PR TITLE
Updated compiler and fix Yojson deprecation warnings.

### PR DESCRIPTION
**Pull request by idkjs**
Studying your code so did some touching up if you want it.

### CODE REVIEW

File: decoders-bencode.opam [hunk 1]
```diff
@@ -11,15 +11,15 @@ doc: "https://mattjbray.github.io/ocaml-decoders/"
 bug-reports: "https://github.com/mattjbray/ocaml-decoders/issues"
 depends: [
   "ocaml" {>= "4.03.0"}
-  "dune" {>= "2.0"}
+  "dune" {>= "2.8" & >= "2.0"}
```
Comment (c-cube) [inline decoders-bencode.opam:14]: that seems redundant, and why bump the minimal version of dune?

Comment (idkjs) [reply decoders-bencode.opam:14]: That change was auto generated by running `dune upgrade`. Revert it?

Comment (idkjs) [reply decoders-bencode.opam:14]: Also, if you ever decide to compile it all with melange, the minimum version is 2.8. That will also get you the `let*` syntax available in bucklescript/rescript when its imported as a dep. If that is of any interest to you.

Comment (c-cube) [reply decoders-bencode.opam:14]: personally I'd revert it, yes. If you use melange, you can add `dune >= 2.8` in your project, it doesn't mean each dependency has to have the same constraint :). Opam will account for them all.

Comment (idkjs) [reply decoders-bencode.opam]: Apparently it was changing the dune version to `2.8` which automatically add the version to the dune files. So reverter `decoders.opam` dune version to what you had.

**Comment by mattjbray** [comment]
Probably about time we moved to `Yojson.t`. If you can just add a constraint on `yojson` to the version that added `Yojson.Basic.t`, etc. to the `dune-project` file here: https://github.com/mattjbray/ocaml-decoders/blob/e00bad1d2e4c2b1394aee9ae5a7a6fe3ab04ecec/dune-project#L117

Then run `dune build` to re-generate the `opam` files, I'll be happy to merge.

**Pull request merged**